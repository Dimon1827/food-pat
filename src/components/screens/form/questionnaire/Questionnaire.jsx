import Information from './information/Information';
import Task from './task/Task';
import fieldCatInformation from './information/fieldCatInformation';
import contactsData from './information/contactsData';
import Commentary from './commentary/Commentary';
import Additionally from './additionally/Additionally';
import './Questionnaire.scss';

function Questionnaire() {
  return (
    <section className="questionnaire">
      <form
        className="form"
        action="https://echo.htmlacademy.ru/"
        method="post"
      >
        <div className="container">
          <div className="questionnaire__wrapper">
            <h1 className="questionnaire__title title">Подбор программы</h1>
            <p className="questionnaire__description">
              Заполните анкету, и мы подберем программу питания для вас
            </p>
          </div>
          <div className="questionnaire__form-wrapper">
            <fieldset className="questionnaire__field information line">
              <legend className="visually-hidden">
                Заполнение информации о коте
              </legend>
              <Information fieldCatInformation={fieldCatInformation} />
            </fieldset>
            <Task />
          </div>
        </div>
        <div className="container">
          <div className="questionnaire__field information">
            <h2 className="information__title form-title">Контактные данные</h2>
            <div className="questionnaire__form-wrapper">
              <Information contactsData={contactsData} />
            </div>
          </div>
        </div>
        <Commentary />
        <Additionally />
      </form>
    </section>
  );
}

export default Questionnaire;
