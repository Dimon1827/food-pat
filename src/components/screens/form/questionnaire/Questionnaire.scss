@import '@/assets/styles/variables';

.questionnaire__description {
  color: $black-color;
  font-family: $oswald-font;
  font-size: 14px;
  line-height: 18px;
  text-transform: uppercase;
  margin: 0;
  margin-bottom: 39px;
  max-width: 226px;

  @media (min-width: $tablet-width) {
    margin-bottom: 66px;
    font-size: 20px;
    line-height: 30px;
    max-width: 323px;
  }
}

.questionnaire__wrapper {
  .questionnaire__title {
    margin-bottom: 37px;
    position: relative;

    @media (max-width: $mobile-width) {
      padding-top: 29px;

      &::before {
        content: "";
        position: absolute;
        top: 0;
        left: -20px;
        width: 106%;
        height: 1px;
        background-color: $special-grey;
      }
    }

    @media (min-width: $tablet-width) {
      margin-bottom: 54px;
      border: none;
    }
  }

  @media (min-width: $tablet-width) and (max-width: $desktop-width) {
    padding-left: 40px;
  }
}

.questionnaire__form-wrapper {
  width: 100%;

  @media (min-width: $desktop-width) {
    display: grid;
    grid-template-columns: repeat(2, 1fr);
    gap: 80px;
    margin-bottom: 69px;
  }
}

.questionnaire__field {
  border: none;
  margin: 0;
  padding: 0;
  position: relative;

  @media (min-width: $tablet-width) {
    padding: 0 40px;
    margin-bottom: 15px;
  }

  @media (min-width: $desktop-width) {
    padding: 0;
  }

  .questionnaire__form-wrapper {
    @media (min-width: $desktop-width) {
      display: flex;
      align-items: center;
    }
  }
}

.questionnaire__button {
  margin-bottom: 21px;
  margin-top: 14px;
  width: 100%;

  @media (min-width: $tablet-width) {
    margin: 0 55px 64px 0;
    font-size: 20px;
    line-height: 26px;
    width: 284px;
  }

  @media (min-width: $desktop-width) {
    width: 570px;
    margin-right: 75px;
  }
}

.questionnaire__text {
  color: $black-color;
  text-align: center;
  line-height: 16px;
  margin: 0;
  margin-bottom: 32px;

  @media (min-width: $tablet-width) {
    margin: 0;
    display: inline-block;
    text-align: start;
  }

  @media (min-width: $desktop-width) {
    align-self: center;
  }
}
