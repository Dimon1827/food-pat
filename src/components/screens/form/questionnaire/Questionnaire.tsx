import { useForm, FormProvider } from 'react-hook-form';
import Information from './information/Information';
import Task from './task/Tasks';
import fieldPersonInformation from './information/fieldPersonInformation';
import contactsData from './information/contactsData';
import Commentary from './commentary/Commentary';
import Additionally from './additionally/Additionally';
import './Questionnaire.scss';
import { Alert, Snackbar } from '@mui/material';
import { useState } from 'react';

function Questionnaire() {

  const [isOpen, setIsOpen] = useState(false);

  const methods = useForm({
    mode: 'onChange' 
  });

  const onSubmit = (data:any) => {
    console.log(data);
    setIsOpen(true);
  };

  return (
    <section className="questionnaire" data-testid="questionnaire">
      <FormProvider {...methods}>
        <form className="form" onSubmit={methods.handleSubmit(onSubmit)} data-testid="form">
          <div className="container">
            <div className="questionnaire__wrapper">
              <h1 className="questionnaire__title title">Подбор программы</h1>
              <p className="questionnaire__description">
                Заполните анкету, и мы подберем программу питания для вас
              </p>
            </div>
            <div className="questionnaire__form-wrapper">
              <fieldset className="questionnaire__field information line">
                <legend className="visually-hidden">
                  Заполнение информации о вас
                </legend>
                <Information fieldPersonInformation={fieldPersonInformation} />
              </fieldset>
              <Task />
            </div>
          </div>
          <div className="container">
            <div className="questionnaire__field information">
              <h2 className="information__title form-title">
                Контактные данные
              </h2>
              <div className="questionnaire__form-wrapper">
                <Information contactsData={contactsData} />
              </div>
            </div>
          </div>
          <Commentary />
          <Additionally />
        </form>
      </FormProvider>
      <Snackbar
        open={isOpen}
        autoHideDuration={100000}
        onClose={() => setIsOpen(false)}
      >
        <Alert severity="success" variant="filled">
          Твои данные успешно отправлены
        </Alert>
      </Snackbar>
    </section>
  );
}

export default Questionnaire;
