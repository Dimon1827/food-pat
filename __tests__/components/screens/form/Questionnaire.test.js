import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import Questionnaire from '@/components/screens/form/questionnaire/Questionnaire';
import React from 'react';

jest.mock('react-hook-form', () => ({
  ...jest.requireActual('react-hook-form'),
  useFormContext: () => ({
    register: jest.fn(),
    handleSubmit: jest.fn((callback) => callback),
    formState: { errors: {} },
  }),
}));

describe('Questionnaire', () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('Questionnaire should render correctly', () => {
    render(<Questionnaire />);
    const informationItems = screen.getAllByTestId('information-item');
    const telephoneIcon = screen.getByTestId('telephone');
    const emailIcon = screen.getByTestId('email');
    const questionnaire = screen.getByTestId('questionnaire');
    expect(questionnaire).toMatchSnapshot();
    expect(informationItems.length).toBe(5);
    expect(telephoneIcon).toBeInTheDocument();
    expect(emailIcon).toBeInTheDocument();
  });
  it('The form should be submitted if all the information has been entered', async () => {
    const { getByTestId } = render(<Questionnaire />);
    const form = getByTestId('form');
    const name = screen.getByPlaceholderText('Елена');
    const weight = screen.getByPlaceholderText('70');
    const age = screen.getByPlaceholderText('35');
    const radioButton = screen.getByDisplayValue('slimming');
    const email = screen.getByPlaceholderText('a.drogalov2015@yandex.ru');
    const telephone = screen.getByPlaceholderText('8 (909) 512-34-21');
    fireEvent.change(name, { target: { value: 'Антон' } });
    fireEvent.blur(name);
    fireEvent.change(weight, { target: { value: '50' } });
    fireEvent.blur(weight);
    fireEvent.click(radioButton, { target: { checked: true } });
    fireEvent.change(age, { target: { value: '35' } });
    fireEvent.blur(age);
    fireEvent.change(email, { target: { value: 'a.drogalov2015@yandex.ru' } });
    fireEvent.blur(email);
    fireEvent.change(telephone, {
      target: { value: '89516361542' },
    });
    fireEvent.blur(telephone);
    fireEvent.submit(form);

    await waitFor(() => {
      const snackbar = screen.getByRole('alert');
      expect(snackbar).toBeInTheDocument();
      expect(snackbar).toHaveTextContent('Твои данные успешно отправлены');
    });
  });
  it('should call setIsOpen(false) when Snackbar is closed', async () => {
    const setIsOpen = jest.fn();
    jest.spyOn(React, 'useState').mockReturnValue([true, setIsOpen]);
    render(<Questionnaire />);

    const form = screen.getByTestId('form');
    fireEvent.submit(form);

    await waitFor(() => {
      const snackbar = screen.getByRole('alert');
      expect(snackbar).toBeInTheDocument();
    });

    const snackbar = screen.getByRole('alert');
    fireEvent.keyDown(snackbar, { key: 'Escape', code: 'Escape' });

    await waitFor(() => {
      expect(setIsOpen).toHaveBeenCalledWith(false);
    });
  });
});
