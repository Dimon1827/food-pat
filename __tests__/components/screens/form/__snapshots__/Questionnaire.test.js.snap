// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Questionnaire Questionnaire should render correctly 1`] = `
<section
  class="questionnaire"
  data-testid="questionnaire"
>
  <form
    class="form"
    data-testid="form"
  >
    <div
      class="container"
    >
      <div
        class="questionnaire__wrapper"
      >
        <h1
          class="questionnaire__title title"
        >
          Подбор программы
        </h1>
        <p
          class="questionnaire__description"
        >
          Заполните анкету, и мы подберем программу питания для вас
        </p>
      </div>
      <div
        class="questionnaire__form-wrapper"
      >
        <fieldset
          class="questionnaire__field information line"
        >
          <legend
            class="visually-hidden"
          >
            Заполнение информации о вас
          </legend>
          <div
            class="information__inner"
            data-testid="information-item"
          >
            <label
              class="information__label label"
              for="name"
            >
              ИМЯ:*
            </label>
            <input
              class="text-input text-input--error"
              id="name"
              placeholder="Елена"
            />
          </div>
          <div
            class="information__inner"
            data-testid="information-item"
          >
            <label
              class="information__label label"
              for="weight"
            >
              Вес (кг):*
            </label>
            <input
              class="text-input text-input--error"
              id="weight"
              placeholder="70"
            />
          </div>
          <div
            class="information__inner"
            data-testid="information-item"
          >
            <label
              class="information__label label"
              for="age"
            >
              Возраст (лет):
            </label>
            <input
              class="text-input "
              id="age"
              placeholder="35"
            />
          </div>
        </fieldset>
        <fieldset
          class="questionnaire__field task"
        >
          <legend
            class="visually-hidden"
          >
            Основная задача
          </legend>
          <ul
            class="task__list line"
          >
            <li
              class="task__item"
            >
              <input
                class="visually-hidden control-input task__input"
                id="slimming"
                name="task"
                type="radio"
                value="slimming"
              />
              <label
                class="label task__label"
                for="slimming"
              >
                Похудение
              </label>
            </li>
            <li
              class="task__item"
            >
              <input
                class="visually-hidden control-input task__input"
                id="gain"
                name="task"
                type="radio"
                value="gain"
              />
              <label
                class="label task__label"
                for="gain"
              >
                Набор массы
              </label>
            </li>
            <li
              class="task__item"
            >
              <input
                class="visually-hidden control-input task__input"
                id="fit"
                name="task"
                type="radio"
                value="fit"
              />
              <label
                class="label task__label"
                for="fit"
              >
                Поддержка формы
              </label>
            </li>
          </ul>
        </fieldset>
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="questionnaire__field information"
      >
        <h2
          class="information__title form-title"
        >
          Контактные данные
        </h2>
        <div
          class="questionnaire__form-wrapper"
        >
          <div
            class="information__inner"
            data-testid="information-item"
          >
            <label
              class="information__label label"
              for="email"
            >
              E-mail:*
            </label>
            <input
              class="text-input text-input--error"
              id="email"
              placeholder="a.drogalov2015@yandex.ru"
            />
            <svg
              class="information__icon"
              data-testid="email"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 1024 1024"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M928 160H96c-17.7 0-32 14.3-32 32v640c0 17.7 14.3 32 32 32h832c17.7 0 32-14.3 32-32V192c0-17.7-14.3-32-32-32zm-40 110.8V792H136V270.8l-27.6-21.5 39.3-50.5 42.8 33.3h643.1l42.8-33.3 39.3 50.5-27.7 21.5zM833.6 232L512 482 190.4 232l-42.8-33.3-39.3 50.5 27.6 21.5 341.6 265.6a55.99 55.99 0 0 0 68.7 0L888 270.8l27.6-21.5-39.3-50.5-42.7 33.2z"
              />
            </svg>
          </div>
          <div
            class="information__inner"
            data-testid="information-item"
          >
            <label
              class="information__label label"
              for="telephone"
            >
              Телефон:*
            </label>
            <input
              class="text-input text-input--error"
              id="telephone"
              placeholder="8 (909) 512-34-21"
            />
            <svg
              class="information__icon"
              data-testid="telephone"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M3.654 1.328a.678.678 0 0 0-1.015-.063L1.605 2.3c-.483.484-.661 1.169-.45 1.77a17.6 17.6 0 0 0 4.168 6.608 17.6 17.6 0 0 0 6.608 4.168c.601.211 1.286.033 1.77-.45l1.034-1.034a.678.678 0 0 0-.063-1.015l-2.307-1.794a.68.68 0 0 0-.58-.122l-2.19.547a1.75 1.75 0 0 1-1.657-.459L5.482 8.062a1.75 1.75 0 0 1-.46-1.657l.548-2.19a.68.68 0 0 0-.122-.58zM1.884.511a1.745 1.745 0 0 1 2.612.163L6.29 2.98c.329.423.445.974.315 1.494l-.547 2.19a.68.68 0 0 0 .178.643l2.457 2.457a.68.68 0 0 0 .644.178l2.189-.547a1.75 1.75 0 0 1 1.494.315l2.306 1.794c.829.645.905 1.87.163 2.611l-1.034 1.034c-.74.74-1.846 1.065-2.877.702a18.6 18.6 0 0 1-7.01-4.42 18.6 18.6 0 0 1-4.42-7.009c-.362-1.03-.037-2.137.703-2.877zm10.762.135a.5.5 0 0 1 .708 0l2.5 2.5a.5.5 0 0 1 0 .708l-2.5 2.5a.5.5 0 0 1-.708-.708L14.293 4H9.5a.5.5 0 0 1 0-1h4.793l-1.647-1.646a.5.5 0 0 1 0-.708"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="questionnaire__field commentary"
        data-testid="commentary"
      >
        <h2
          class="commentary__title form-title"
        >
          Комментарий
        </h2>
        <textarea
          class="commentary__text"
          cols="30"
          placeholder="Расскажите о вас. Что предпочитаете и т.д"
          rows="10"
        />
        <span
          class="commentary__counter"
        >
          Осталось букв: 
          200
        </span>
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="questionnaire__field additionally"
        data-testid="additionally"
      >
        <h2
          class="additionally__title form-title"
        >
          Аллергены
        </h2>
        <ul
          class="additionally__wrapper"
        >
          <li
            class="additionally__item"
          >
            <input
              class="visually-hidden control-input additionally__input"
              id="milk"
              name="milk"
              type="checkbox"
              value="milk"
            />
            <label
              class="label additionally__label"
              for="milk"
            >
              Молоко
            </label>
          </li>
          <li
            class="additionally__item"
          >
            <input
              class="visually-hidden control-input additionally__input"
              id="fish"
              name="fish"
              type="checkbox"
              value="fish"
            />
            <label
              class="label additionally__label"
              for="fish"
            >
              Рыба
            </label>
          </li>
          <li
            class="additionally__item"
          >
            <input
              class="visually-hidden control-input additionally__input"
              id="nuts"
              name="nuts"
              type="checkbox"
              value="nuts"
            />
            <label
              class="label additionally__label"
              for="nuts"
            >
              Орехи
            </label>
          </li>
          <li
            class="additionally__item"
          >
            <input
              class="visually-hidden control-input additionally__input"
              id="egg"
              name="egg"
              type="checkbox"
              value="egg"
            />
            <label
              class="label additionally__label"
              for="egg"
            >
              Яйца
            </label>
          </li>
        </ul>
        <button
          class="questionnaire__button button"
          type="submit"
        >
          отправить заявку
        </button>
        <p
          class="questionnaire__text"
        >
          * — Обязательные поля
        </p>
      </div>
    </div>
  </form>
</section>
`;
